sql/core/src/main/scala/org/apache/spark/sql/json/JSONRelation.scala

```diff
@@ -101,32 +103,87 @@ private[sql] class DefaultSource
   }
 }
 
-private[sql] case class JSONRelation(
-    path: String,
-    samplingRatio: Double,
+private[sql] class JSONRelation(
+    // baseRDD is not immutable with respect to INSERT OVERWRITE
+    // and so it must be recreated at least as often as the
+    // underlying inputs are modified. To be safe, a call-by-name
+    // value (a function) is used instead of a regular value to
+    // ensure the RDD is recreated on each and every operation.
+    baseRDD: => RDD[String],
```

Comment:
@NathanHowell Do you think a closure at here will be better (as mentioned in https://github.com/databricks/scala-style-guide#call_by_name)?